server/app/services/scheduler/strategy/daemon.rb

```diff
@@ -15,11 +15,33 @@ def host_grace_period
         10.minutes
       end
 
+      # @param [GridService] grid_service
+      # @param [Integer] instance_number
+      # @param [Array<Scheduler::Node>] nodes
+      # @return [Scheduler::Node,NilClass]
+      def find_stateless_node(grid_service, instance_number, nodes)
+        prev_instance = grid_service.grid_service_instances.has_node.find_by(
+          instance_number: instance_number
+        )
+        if prev_instance
+          node = nodes.find { |n| n.node == prev_instance.host_node }
+          unless node
+            selector = (instance_number.to_f / nodes.size.to_f).floor
+            candidates = self.sort_candidates(nodes, grid_service, instance_number)
+            node = candidates.select { |c| c.schedule_counter <= selector }.last
```

Comment:
I'm not sure I understand what this is doing vs the `self.sort_candidates(nodes, grid_service, instance_number).first` - the `sort_candidates` already takes the `node.schedule_counter` into account, so that should end up picking the node with the fewest scheduled instances?

Reply:
Updated description.. also specs should demonstrate how this behaves.

Reply:
Right, so the `.last` here is to effectively select those nodes that were running the high-numbered instances that are no longer part of the service after some hosts went offline. However, I'm a little concerned about this being a bit fragile... I think this assumes that e.g. the `sort_candidates` ranking of the nodes is stable across the scheduler runs... if the ranks change to put some host node with a lower-numbered instance at the end, then could you actually end up scheduling multiple instances onto that same node, while leaving the node that used to be running the high-numbered instances without any scheduled instances? It would need some actual testing to figure out if there are some kind of nasty corner cases like that...

I'm not sure how to make it less fragile either... perhaps it should explicitly select the `grid_service_instances.where(:instance_number.gt => total_instances)` nodes, but that's hard to fit in with the actual scheduler interface and the way it's used, particularly because the `GridServiceDeployer` deletes the high-numbered instances before actually scheduling the nodes.

And to answer my own question: if this was just `self.sort_candidates(...).first`, then it would end up scheduling multiple instances onto the same low-numbered nodes that already have instances running:

```
  1) Scheduler::Strategy::Daemon#find_node with existing nodes minimizes shuffle when nodes 2 & 3 are missing
     Failure/Error: expect(scheduled.map {|s| s.node_number}).to eq([
       
       expected: [1, 10, 9, 4, 5, 6, 7, 8]
            got: [1, 4, 5, 4, 5, 6, 7, 8]
```

Reply:
Oh NVM, the `daemon` strategy has its own `sort_candidates` which looks like it should be perfectly stable. I was looking at the `ha` strategy's `sort_candidates`.